source/common/access_log/access_log_formatter.cc

```diff
@@ -326,195 +330,398 @@ std::vector<FormatterProviderPtr> AccessLogFormatParser::parse(const std::string
+    return stringValue(field_extractor_(stream_info));
+  }
+
+private:
+  FieldExtractor field_extractor_;
+};
+
+// StreamInfo absl::optional<std::string> field extractor.
+class StreamInfoOptionalStringFieldExtractor : public StreamInfoFormatter::FieldExtractor {
+public:
+  using FieldExtractor = std::function<absl::optional<std::string>(const StreamInfo::StreamInfo&)>;
+
+  StreamInfoOptionalStringFieldExtractor(FieldExtractor f) : field_extractor_(f) {}
+
+  // StreamInfoFormatter::FieldExtractor
+  std::string extract(const StreamInfo::StreamInfo& stream_info) const override {
+    auto str = field_extractor_(stream_info);
+    if (!str) {
+      return UnspecifiedValueString;
+    }
+
+    return str.value();
+  }
+  ProtobufWkt::Value extractValue(const StreamInfo::StreamInfo& stream_info) const override {
+    auto str = field_extractor_(stream_info);
+    if (!str) {
+      return unspecifiedValue();
+    }
+
+    return stringValue(str.value());
+  }
+
+private:
+  FieldExtractor field_extractor_;
+};
+
+// StreamInfo std::chrono_nanoseconds field extractor.
+class StreamInfoDurationFieldExtractor : public StreamInfoFormatter::FieldExtractor {
+public:
+  using FieldExtractor =
+      std::function<absl::optional<std::chrono::nanoseconds>(const StreamInfo::StreamInfo&)>;
+
+  StreamInfoDurationFieldExtractor(FieldExtractor f) : field_extractor_(f) {}
 
+  // StreamInfoFormatter::FieldExtractor
+  std::string extract(const StreamInfo::StreamInfo& stream_info) const override {
+    auto millis = extractMillis(stream_info);
+    if (!millis) {
+      return UnspecifiedValueString;
+    }
+
+    return fmt::format_int(millis.value()).str();
+  }
+  ProtobufWkt::Value extractValue(const StreamInfo::StreamInfo& stream_info) const override {
+    auto millis = extractMillis(stream_info);
+    if (!millis) {
+      return unspecifiedValue();
+    }
+
+    return numberValue(millis.value());
+  }
+
+private:
+  absl::optional<uint32_t> extractMillis(const StreamInfo::StreamInfo& stream_info) const {
+    auto time = field_extractor_(stream_info);
+    if (time) {
+      return std::chrono::duration_cast<std::chrono::milliseconds>(time.value()).count();
+    }
+    return absl::nullopt;
+  }
+
+  FieldExtractor field_extractor_;
+};
+
+// StreamInfo uint64_t field extractor.
+class StreamInfoUInt64FieldExtractor : public StreamInfoFormatter::FieldExtractor {
+public:
+  using FieldExtractor = std::function<uint64_t(const StreamInfo::StreamInfo&)>;
+
+  StreamInfoUInt64FieldExtractor(FieldExtractor f) : field_extractor_(f) {}
+
+  // StreamInfoFormatter::FieldExtractor
+  std::string extract(const StreamInfo::StreamInfo& stream_info) const override {
+    return fmt::format_int(field_extractor_(stream_info)).str();
+  }
+  ProtobufWkt::Value extractValue(const StreamInfo::StreamInfo& stream_info) const override {
+    return numberValue(field_extractor_(stream_info));
```

Comment:
Good to know it is fine to pack `uint64_t` as a `double`.

Reply:
yeah unless the value is expected to be >= 2^53.

Reply:
My concern was more about when it is serialized (whether or not the serializer adds `.0`).

Reply:
by JSON spec it has only double as number value, so whether it adds `.0` doesn't matter.

Reply:
I'll add some docs and include a mention about large integer values.